docs/user/rstudio/ide/guide/tools/copilot.qmd

```diff
@@ -0,0 +1,237 @@
+:::
+
+### Comments
+
+When using Copilot to solve a broad problem, it is a good idea to provide a high-level goal at the beginning of the file. This will help Copilot provide suggestions that are relevant to the specific problem you are trying to solve.
+
+``` r
+# This script will do x
+# using the packages x,y,z
+# other constraints or details
+```
+
+A real example might be:
+
+``` r
+# Create a robust function to calculate the standard error of a vector
+# The function name will be calc_se with an argument x for the vector, with 
+# an argument to remove missing values called na.rm, and a default value of TRUE
+# The function should have type checking to only allow numeric vectors
+```
+
+Which might generate something like the below:
+
+``` r
+calc_se <- function(x, na.rm = TRUE) {
+  if (!is.numeric(x)) {
+    stop("x must be numeric")
+  }
+  if (na.rm) {
+    x <- x[!is.na(x)]
+  }
+  sqrt(var(x) / length(x))
+}
+```
+
+Which generated a working example!
+
+``` r
+calc_se(1:10)
+#> [1] 0.9574
+
+# note the function also properly indicates the need for a numeric input
+calc_se(letters)
+#> Error in calc_se(letters) : x must be numeric
+```
+
+For smaller problems or to provide context or guidance within a function or code block, comments can be used to provide a basic prompt for Copilot.
+
+Comments in code such as:
+
+``` r
+# create a function that takes a number and returns the square
+```
+
+Might be used by Copilot to suggest a function that takes a number and returns the square:
+
+``` r
+# create a function that takes a number and returns the square
+square_number <- function(x) {
+  x^2
+}
+```
+
+Lastly, comments can be used inline within code to provide additional context or guidance to Copilot. For example:
+
+``` r
+library(dplyr)
+
+mtcars |> 
+   group_by(cyl) |> 
+   summarize(
+      mean = mean(mpg),
+      # add the standard error for mpg
+   )
+```
+
+Might return:
+
+``` r
+mtcars |> 
+   group_by(cyl) |> 
+   summarize(
+      mean = mean(mpg),
+      # add the standard error for mpg
+      se = sd(mpg) / sqrt(n())
+   )
+```
+
+However, the comment may not needed if a descriptive column name is used instead.
+
+![](images/copilot_se-inline.png){fig-alt="Copilot inline completion"}
+
+### Language support
+
+While R is the primary expected language for RStudio users, Copilot also supports other languages such as SQL, Python, HTML, CSS, JavaScript, and many other languages. Copilot will use the file extension or other context to determine the language of the active document. In multi-language Quarto documents for example, Copilot could use the active language within a code chunk or the body of the document to generate suggestions.
+
+Comments in other languages can be used to generate suggestions in the active document For example, a HTML comment (`<!-- COMMENT BODY -->`) might be used in the body of a `.qmd` file:
```

Comment:
Missing period: Change
   the active document For example
to
   the active document. For example

Reply:
Thx!